---
title: "ASXL1 Mutant, 6 month pilot analysis"
author: "Theresa Lusardi"
date: "2019.05.01"
output:
  html_document:
    df_print: paged
    toc: yes
  html_notebook:
    toc: yes
    toc_float: yes
---

Initial analysis of ASXL1 mutant vs Wildtype mouse bone marrow (n = 3 / group)
Bone marrow was isolated from male littermate mice on a C57Bl6 background
Enriched for lin- cells
Each mouse was labeled with Biolegend hashtag anti-CD45

Fastq data is archived at:  /home/groups/CEDAR/archive/seq/CEL190322TB/190426_D00735_0326_ACDLYRANXX
New Fastq (with barcodes) archived at:
  /home/groups/CEDAR/archive/seq/CEL190322TB_new/190426_D00735_0326_ACDLYRANXX/Unaligned
Analysis files at: /home/groups/CEDAR/lusardi/asxl/sixmo


```{r init, echo=FALSE}
# Reticulate library needed to use umap-learn
library(reticulate)
use_condaenv(condaenv="r-reticulate", conda='/Users/lusardi/anaconda3/bin/conda')
library(dplyr)
library(Seurat)

# Load the dataset
data_dir <- "/Users/lusardi/Documents/CEDAR/Projects/3.ASXL_mutant/sixMonthRun/data/lane1/raw_feature_bc_matrix/"
filt_dir <- "/Users/lusardi/Documents/CEDAR/Projects/3.ASXL_mutant/sixMonthRun/data/lane1/filtered_feature_bc_matrix"
lane1.data <- Read10X(data.dir = filt_dir)
lane1.gex = lane1.data$`Gene Expression`
lane1.ab = lane1.data$`Antibody Capture`
#lane2.data <- Read10X(data.dir = "")
#lane2.gex = lane2.data$`Gene Expression`

# Initialize the Seurat object with the raw (non-normalized data).
lane1 <- CreateSeuratObject(counts = lane1.gex, project = "lane1_6mo", min.cells = 3, min.features = 200)
lane1
```

Take a look at the expression of the genes that were used for antibody labeling
```{r, echo=FALSE}
# Flow Genes corresponding to antibody capture antigens
# Note the list names correspond to the identifier used for the Ab reference, the list elements are the corresponding gene IDs from MGI.
ab_gene = list(ab_CD117 = c("Kit"),
               ab_CD150 = c("Slamf1"),
               ab_CD16_32 = c("Fcgr2b", "Fcgr3", "Fcgr4"),
               ab_CD48 = c("Cd48"),
               ab_Sca1 = c("Ly6a", "Ly6e"),
               ab_CD41 = c("Itga2b"),
               ab_CD105 = c("Eng"))
genelist = unlist(ab_gene)
lane1.gex[genelist, 1:50]
```

Calculate the percent mitochondrial genes for each cell - %mito genes > 5 indicates cells are dead/dying
```{r mito, fig.height=7, fig.width=13, echo=FALSE}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
lane1[["percent.mt"]] <- PercentageFeatureSet(object = lane1, pattern = "^mt-")
```


```{r qc, fig.height=7, fig.width=13, echo=FALSE}
# Show QC metrics for the first 5 cells
head(x = lane1@meta.data, 15)
```

```{r qc2, fig.height=7, fig.width=13, echo=FALSE}

#Visualize QC metrics as a violin plot
VlnPlot(object = lane1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(object = lane1, feature1 = "nCount_RNA", feature2 = "percent.mt") 
plot2 <- FeatureScatter(object = lane1, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") 
CombinePlots(plots = list(plot1,plot2))

```

Based on plot values subset the data to filter out outlier cells - cut at 10%
```{r, echo=FALSE}
lane1 <- subset(x = lane1, subset = nFeature_RNA > 100 & nFeature_RNA < 4000 & percent.mt < 10)
```

### Normalizing the data
Default Seurat normalization scheme.
After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method "LogNormalize" that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. 

```{r normalize, echo=FALSE}
lane1 <- NormalizeData(object = lane1, normalization.method = "LogNormalize", scale.factor = 1e4)
```

### Identification of highly variable features (feature selection)
```{r var_features, fig.height=5, fig.width=11, echo=FALSE}
lane1 <- FindVariableFeatures(object = lane1,selection.method = 'vst', nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(x = VariableFeatures(object = lane1), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(object = lane1)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
CombinePlots(plots = list(plot1, plot2))
```

***

### Scaling the data
Apply a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The `ScaleData` function:

* Shifts the expression of each gene, so that the mean expression across cells is 0
* Scales the expression of each gene, so that the variance across cells is 1
    + This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
* The results of this are stored in `lane1[["RNA"]]@scale.data`

```{r regress, fig.height=7, fig.width=11, results='hide', echo=FALSE}
all.genes <- rownames(x = lane1)
lane1 <- ScaleData(object = lane1, features = all.genes)
```

***

### Perform linear dimensional reduction

Next we perform PCA on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using `features` argument if you wish to choose a different subset.

```{r pca,results='hide', echo=FALSE}
lane1 <- RunPCA(object = lane1, features = VariableFeatures(object = lane1))
```

Seurat provides several useful ways of visualizing both cells and features that define the PCA, including `VizDimReduction`, `DimPlot`, and `DimHeatmap`

```{r pca_viz, message=TRUE, echo=FALSE}
# Examine and visualize PCA results a few different ways
print(x = lane1[['pca']], dims = 1:5, nfeatures = 5)
VizDimLoadings(object = lane1, dims = 1:2, reduction = 'pca')
DimPlot(object = lane1, reduction = 'pca')
```

In particular `DimHeatmap` allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. Setting `cells` to a number plots the 'extreme' cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated feature sets.

```{r single-heatmap, echo=FALSE}
DimHeatmap(object = lane1, dims = 1, cells = 500, balanced = TRUE)
```

```{r multi-heatmap, fig.height=12, fig.width=9, echo=FALSE}
DimHeatmap(object = lane1, dims = 1:20, cells = 500, balanced = TRUE)
```

### Determine the 'dimensionality' of the dataset

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a 'metafeature' that combines information across a correlated feature set. The top principal components therefore represent a robust compression of the dataset. However, how many componenets should we choose to include? 10? 20? 100?

In [Macosko *et al*](http://www.cell.com/abstract/S0092-8674(15)00549-8), we implemented a resampling test inspired by the JackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a 'null distribution' of feature scores, and repeat this procedure. We identify 'significant' PCs as those who have a strong enrichment of low p-value features.

```{r jackstraw, fig.height=6, fig.width=10, echo=FALSE}
# NOTE: This process can take a long time for big datasets, comment out for expediency. More approximate techniques such as those implemented in ElbowPlot() can be used to reduce computation time
lane1 <- JackStraw(object = lane1, num.replicate = 100)
lane1 <- ScoreJackStraw(object = lane1, dims = 1:20)
```

The `JackStrawPlot` function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). 'Significant' PCs will show a strong enrichment of features with low p-values (solid curve above the dashed line). In this case it appears that there is a sharp drop-off in significance after the first 10-12 PCs.

```{r jsplots, fig.height=6, fig.width=10, echo=FALSE}
JackStrawPlot(object = lane1, dims = 1:20)
```

An alternative heuristic method generates an 'Elbow plot': a ranking of principle components based on the percentage of variance explained by each one (`ElbowPlot` function). In this example, we can observe an 'elbow' around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs. 

```{r elbow_plot, fig.height=6, fig.width=10, echo=FALSE}
ElbowPlot(object = lane1)
```

### Cluster the cells

Seurat v3 applies a graph-based clustering approach, building upon initial strategies in ([Macosko *et al*](http://www.cell.com/abstract/S0092-8674(15)00549-8)). Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [[SNN-Cliq, Xu and Su, Bioinformatics, 2015]](http://bioinformatics.oxfordjournals.org/content/early/2015/02/10/bioinformatics.btv088.abstract) and CyTOF data [[PhenoGraph, Levine *et al*., Cell, 2015]](http://www.ncbi.nlm.nih.gov/pubmed/26095251). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected 'quasi-cliques' or 'communities'. 

As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the `FindNeighbors` function, and takes as input the previously defined dimensionality of the dataset (first 10 PCs).

To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM [[SLM, Blondel *et al*., Journal of Statistical Mechanics]](http://dx.doi.org/10.1088/1742-5468/2008/10/P10008), to iteratively group cells together, with the goal of optimizing the standard modularity function. The `FindClusters` function implements this procedure, and contains a resolution parameter that sets the 'granularity' of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters can be found using the `Idents` function.


```{r cluster, fig.height=5, fig.width=7, echo=FALSE}
lane1 <- FindNeighbors(object = lane1, dims = 1:10)
lane1 <- FindClusters(object = lane1, resolution = 0.5)

# Look at cluster IDs of the first 5 cells
head(x = Idents(object = lane1), 5)
```

***

### Run non-linear dimensional reduction (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP, to visualize and explore these datasets. The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. As input to the UMAP and tSNE, we suggest using the same PCs as input to the clustering analysis.

```{r tsne, fig.height=5, fig.width=7, echo=FALSE}
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages = "umap-learn")

# TAGL - set up to run UMAP
# If not, go to terminal and figure out what environment is active, and activate (as necessary)
# https://rstudio.github.io/reticulate/articles/versions.html
#reticulate::py_install(packages = "umap-learn", envname = 'r-reticulate')

lane1 <- RunUMAP(object = lane1, dims = 1:20, min.dist = 0.01)
DimPlot(object = lane1, reduction = 'umap', label = TRUE)
lane1 <- RunUMAP(object = lane1, dims = 1:20, min.dist = 0.3)
DimPlot(object = lane1, reduction = 'umap', label = TRUE)
lane1 <- RunUMAP(object = lane1, dims = 1:20, min.dist = 0.5)
DimPlot(object = lane1, reduction = 'umap', label = TRUE)
lane1 <- RunUMAP(object = lane1, dims = 1:20, min.dist = 0.7)
DimPlot(object = lane1, reduction = 'umap', label = TRUE)
```

```{r tsneplot, fig.height=5, fig.width=7, echo=FALSE}
# note that you can set `label = TRUE` or use the LabelClusters function to help label individual clusters
DimPlot(object = lane1, reduction = 'umap', label = TRUE)
```

### Finding differentially expressed features (cluster biomarkers)

Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in `ident.1`), compared to all other cells.  `FindAllMarkers` automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The `min.pct` argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, `max.cells.per.ident` can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed features will likely still rise to the top.

```{r markers1, fig.height=8, fig.width=15, echo=FALSE}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(object = lane1, ident.1 = 1, min.pct = 0.25)
head(x = cluster1.markers, n = 5)
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(object = lane1, ident.1 = 5, ident.2 = c(0, 3), min.pct = 0.25)
head(x = cluster5.markers, n = 5)
# find markers for every cluster compared to all remaining cells, report only the positive ones
lane1.markers <- FindAllMarkers(object = lane1, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
lane1.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
```

Seurat has several tests for differential expression which can be set with the test.use parameter (see our [DE vignette](http://satijalab01.nygenome.org/seurat/v3.0/de_vignette.html) for details). For example, the ROC test returns the 'classification power' for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r markersroc, fig.height=8, fig.width=15}
cluster1.markers <- FindMarkers(object = lane1, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```

We include several tools for visualizing marker expression. `VlnPlot` (shows expression probability distributions across clusters), and `FeaturePlot` (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring `RidgePlot`, `CellScatter`, and `DotPlot` as additional methods to view your dataset.

```{r id_hsc, fig.height=8, fig.width=12, echo=FALSE}
LSK = c(ab_gene$ab_CD117, ab_gene$ab_Sca1)
VlnPlot(object = lane1, features = c(LSK))
FeaturePlot(object = lane1, features = c(LSK))
RidgePlot(object = lane1, features = c(LSK))
DotPlot(object = lane1, features = c(LSK))
```

```{r markerplots, fig.height=8, fig.width=15}
VlnPlot(object = lane1, features = c("Kit", ))
VlnPlot(object = lane1, features = c("Kit", "Slamf1","Ly6a"))
# you can plot raw counts as well
VlnPlot(object = lane1, features = c("Kit", "Slamf1"), slot = 'counts', log = TRUE)
FeaturePlot(object = lane1, features = c("Kit", "Slamf1", "Fcgr2b", "Fcgr3", "Fcgr4", "CD48", "Ly6a", "Ly6e", "Itga2b", "Eng"))
```

`DoHeatmap` generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.

```{r clusterHeatmap, fig.height=8, fig.width=15}
lane1.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC) -> top10
DoHeatmap(object = lane1, features = top10$gene) + NoLegend()
```

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

